pkg/xds/server/callbacks/dataplane_sync_tracker.go

```diff
@@ -89,18 +89,16 @@ func (t *dataplaneSyncTracker) OnStreamRequest(streamID core_xds.StreamID, req u
 		defer t.Unlock()
 		streams := t.dpStreams[dataplaneKey]
 		if streams.activeStreams == nil {
-			streams.activeStreams = map[core_xds.StreamID]bool{}
+			streams.activeStreams = map[core_xds.StreamID]context.CancelFunc{}
 		}
-		streams.activeStreams[streamID] = true
-		if streams.watchdogCancel == nil { // watchdog was not started yet
-			stopCh := make(chan struct{})
-			streams.watchdogCancel = func() {
-				close(stopCh)
-			}
-			// kick off watchdog for that Dataplane
-			go t.newDataplaneWatchdog(proxyId, streamID).Start(stopCh)
-			dataplaneSyncTrackerLog.V(1).Info("started Watchdog for a Dataplane", "streamid", streamID, "proxyId", proxyId, "dataplaneKey", dataplaneKey)
+		stopCh := make(chan struct{})
+		streams.activeStreams[streamID] = func() {
+			close(stopCh)
+			dataplaneSyncTrackerLog.V(1).Info("Stopping active stream for a Dataplane", "streamid", streamID, "proxyId", proxyId, "dataplaneKey", dataplaneKey)
 		}
+		// kick off watchdog for that Dataplane
+		go t.newDataplaneWatchdog(proxyId, streamID).Start(stopCh)
+		dataplaneSyncTrackerLog.V(1).Info("started Watchdog for a Dataplane", "streamid", streamID, "proxyId", proxyId, "dataplaneKey", dataplaneKey)
```

Comment:
Sorry, I don't think this is the way to go.

I see how old streamID in watchdog is a problem, but I think there should be only one watchdog for dataplane.
Additionally, we use the same sync tracker for SDS where this is normal (for now, eventually this will be moved to ADS) to have many streams from one dataplane.

Let me think how we can solve this

Reply:
Yeah, I thought about this too, I was not sure if this was a problem or not.
From what I saw, there was two stream only when I created the overlap using toxiproxy.

I guess a more elegant solution would be to not pass the stream directly and freeze the metadata etc in the watchdog.
What do you think?